<html>
    <head>
    <title>Node JS</title>
    </head>
    <body>
        <h1>Node JS</h1>
        <ul>
            <li>Node.jsÂ® is a JavaScript runtime built on Chrome's V8 JavaScript engine. Node.js uses an event-driven, non-blocking I/O model that makes it lightweight and efficient. Node.js' package ecosystem, npm, is the largest ecosystem of open source libraries in the world.</li>
        </ul>
    <span class="anchor" id="css2"></span>
        <div class="cont" id="css">
            <h3>Assert</h3>
            <ul>
                <li>The assert module provides a simple set of assertion tests that can be used to test invariants.</li>
            <ul>
                <li>Class: assert.AssertionError</li>
                <ul>
                    <li>A subclass of Error that indicates the failure of an assertion.  All errors thrown by the assert module will be instances of the AssertionError class.</li>
                </ul>
                <li>new assert.AssertionError(options)</li>
                <ul>
                    <li>options &lt;Object&gt;</li>
                    <ul>
                        <li>message &lt;string&gt; = error message is going to be set to this value.</li>
                    </ul>
                    <li>actual &lt;any&gt;</li>
                    <ul>
                        <li>actual property on the error instance is going to contain this value.</li>
                    </ul>
                    <li>expected &lt;any&gt;</li>
                    <ul>
                        <li>expected property on the error instance is going to contain this value.</li>
                    </ul>
                    <li>operator &lt;string&gt;</li>
                    <ul>
                        <li>used to indicate what operation was used for comparison.</li>
                    </ul>
                    <li>stackStartFn &lt;function&gt;</li>
                    <ul>
                        <li>if provided, the generated stack trace is going to remove all frames up to the provided function.</li>
                    </ul>
                </ul>
                <li>All instances contain the built-in Error properties (message and name) and:</li>
                <ul>
                    <li>actual &lt;any&gt;</li>
                    <ul>
                        <li>set to the actual value in case</li>
                        <li>Example:</li>
                        <div class="code-sample">
                            <span class="indent"></span>
                            assert.strictEqual()
                        </div>
                    </ul>
                    <li>expected &lt;any&gt;</li>
                    <ul>
                    <li>Set to the expected value in case</li>
                    </ul>
                    <li>generatedMessage &lt;boolean&gt;</li>
                    <ul>
                        <li>Indicates if the message was auto-generated (true) or not.</li>
                    </ul>
                    <li>code &lt;string&gt;</li>
                    <ul>
                        <li>This is always set to the string ERR_ASSERTION to indicate that the error is actually an assertion error.</li>
                    </ul>
                    <li>operator &lt;string&gt;</li>
                    <ul>
                        <li>Set to the passed in operator value.</li>
                    </ul>
                    <span class="anchor" id="sampcode2"></span>
                    <a href="#sampcode2" onclick="toggle_visibility('sampcode1');">Sample Code</a> 
                    <div class="code-sample" id="sampcode1">
                    const assert = require('assert');<br>
                    <span class="comment">// Generate an AssertionError to compare the error message later:</span><br>
                    const { message } = new assert.AssertionError({<br>
                    actual: 1,<br>
                    expected: 2,<br>
                    operator: 'strictEqual'<br>
                    });<br>
                    <span class="comment">// Verify error output:</span><br>
                    try {<br>
                    assert.strictEqual(1, 2);<br>
                    } catch (err) {<br>
                    assert(err instanceof assert.AssertionError);<br>
                    assert.strictEqual(err.message, message);<br>
                    assert.strictEqual(err.name, 'AssertionError<br> [ERR_ASSERTION]');<br>
                    assert.strictEqual(err.actual, 1);<br>
                    assert.strictEqual(err.expected, 2);<br>
                    assert.strictEqual(err.code, 'ERR_ASSERTION');<br>
                    assert.strictEqual(err.operator, 'strictEqual');<br>
                    assert.strictEqual(err.generatedMessage, true);<br>
                    }<br>
                    </div>
                </ul>
                <li>Strict Mode</li>
                <ul>
                    <li>When using the strict mode, any assert function will use the equality used in the strict function mode. So assert.deepEqual() will, for example, work the same as assert.deepStrictEqual().</li>
                    <li>On top of that, error messages which involve objects produce an error diff instead of displaying both objects. That is not the case for the legacy mode.</li>
                    <ul>
                    <div class="code-sample">
                        const assert = require('assert').strict;
                    </div>
                    </ul>
                    <li>Example error diff (the expected, actual, and Lines skipped will be on a single row):</li>
                    <span class="anchor" id="sampcode6"></span>
                    <a href="#sampcode6" onclick="toggle_visibility('sampcode5');">Sample Code</a> 
                    <div class="code-sample" id="sampcode5">
                        const assert = require('assert').strict;
                        assert.deepEqual([[[1, 2, 3]], 4, 5], [[[1, 2, '3']], 4, 5]);
                    </div>
                    <span class="comment">/*AssertionError [ERR_ASSERTION]: Input A expected to deepStrictEqual input B:
                    <ul>
                        + expected<br>
                        - actual<br>
                        ... Lines skipped<br>
                        [<br>
                        [<br>
                        ...<br>
                        2,<br>
                        -     3<br>
                        +     '3'<br>
                        ],<br>
                        ...<br>
                        5<br>
<<<<<<< HEAD
                        ]*/
                    </ul>
                    </span>
                    </ul>
                    <li>Legacy Mode</li>
                    <ul>
                        <li>When accessing assert directly instead of using the strict property, the Abstract Equality Comparison will be used for any function without "strict" in its name, such as assert.deepEqual().</li>
                        <br>
                        <li type = "none">It can be accessed using:</li>
                        <ul>
                        <li type = "none"><div class="code-sample">
                            const assert = require('assert');</div></li>
                        </ul>
                        <li>It is recommended to use the strict mode instead as the Abstract Equality Comparison can often have surprising results. This is especially true for assert.deepEqual(), where the comparison rules are lax:</li>
                        <ul>
                            <li type = "none"><div class="code-sample">
                                <span class="comment"> // WARNING: This does not throw an AssertionError!<br></span>
                                assert.deepEqual(/a/gi, new Date());
                            </div>
                            </li>
                        </ul>
                </ul>
                <li>assert(value[, message])</li>
                <ul>
                    <li>Added in v0.5.9</li>
                    <ul>
                        <li>value &lt;any&gt;</li>
                        <li>message &lt;any&gt;</li>
                    </ul>
                    <li>An alias of assert.ok().</li>
                </ul>
                <li>assert.deepEqual(actual, expected[, message])</li>
                <ul>
                    <table border = "1">
                        <tr>
                            <td>Version</td>
                            <td>Changes</td>
                        </tr>
                        <tr>
                            <td>v9.0.0</td>
                            <td>The Error names and messages are now properly compared</td>
                        </tr>
                        <tr>
                            <td>v8.0.0</td>
                            <td>The Set and Map content is also compared</td>
                        </tr>
                        <tr>
                            <td>v6.4.0, v4.7.1</td>
                            <td>Typed array slices are handled correctly now.</td>
                        </tr>
                        <tr>
                            <td>v6.1.0, v4.5.0</td>
                            <td>Objects with circular references can be used as inputs now.</td>
                        </tr>
                        <tr>
                            <td>v5.10.1, v4.4.3</td>
                            <td>Handle non-Uint8Array typed arrays correctly.</td>
                        </tr>
                        <tr>
                            <td>v0.1.21</td>
                            <td>Added in: v0.1.21</td>
                        </tr>
                    </table>
                    <ul>
                    <li>actual &lt;any&gt;</li>
                    <li>expected&lt;any&gt;</li>
                    <li>message &lt;any&gt;</li>
                    </ul>
                    <li>Strict Mode</li>
                    <ul>
                        <li>An alias of assert.deepStrictEqual().</li>
                    </ul>
                    <li>Legacy Mode</li>
                    <ul>
                        <li>Tests for deep equality between the actual and expected parameters. Primitive values are compared with the Abstract Equality Comparison ( == ).</li>
                        <li>Only enumerable "own" properties are considered. The assert.deepEqual() implementation does not test the [[Prototype]] of objects or enumerable own Symbol properties. For such checks, consider using assert.deepStrictEqual() instead. assert.deepEqual() can have potentially surprising results. The following example does not throw an AssertionError because the properties on the RegExp object are not enumerable:</li>
                        <ul>
                        <div class="code-sample">
                            <span class="comment">// WARNING: This does not throw an AssertionError!</span>  
                            assert.deepEqual(/a/gi, new Date());
                        </div>
                        </ul>
                        <li>An exception is made for Map and Set. Maps and Sets have their contained items compared too, as expected.</li>
                    </ul>
                </ul>
                <li>assert.deepStrictEqual(actual, expected[, message])</li>
                <ul>
                    <table border = "1">
                        <tr>
                            <td>Version</td>
                            <td>Changes</td>
                        </tr>
                        <tr>
                            <td>v9.0.0</td>
                            <td>Enumerable symbol properties are now compared.</td>
                        </tr>
                        <tr>
                            <td>v9.0.0</td>
                            <td>The NaN is now compared using the SameValueZero comparison.</td>
                        </tr>
                        <tr>
                            <td>v8.5.0</td>
                            <td>The Error names and messages are now properly compared</td>
                        </tr>
                        <tr>
                            <td>v8.0.0</td>
                            <td>The Set and Map content is also compared</td>
                        </tr>
                        <tr>
                            <td>v6.4.0, v4.7.1</td>
                            <td>Typed array slices are handled correctly now.</td>
                        </tr>
                        <tr>
                            <td>v6.1.0</td>
                            <td>Objects with circular references can be used as inputs now.</td>
                        </tr>
                        <tr>
                            <td>v5.10.1, v4.4.3</td>
                            <td>Handle non-Uint8Array typed arrays correctly.</td>
                        </tr>
                        <tr>
                            <td>v1.2.0</td>
                            <td>Added in: v1.2.0</td>
                        </tr>
                    </table>
                    <ul>
                    <li>actual &lt;any&gt;</li>
                    <li>expected&lt;any&gt;</li>
                    <li>message &lt;any&gt;</li>
                    </ul>
                    <li>Tests for deep equality between the actual and expected parameters. "Deep" equality means that the enumerable "own" properties of child objects are recursively evaluated also by the following rules.</li>
                    
                </ul>
                <li>Comparison Details</li>
                <ul>
                    <li>Primitive values are compared using the SameValue Comparison, used by Object.is()</li>
                    <li>Type tags of objects should be the same.</li>
                    <li>[[Prototype]] of objects are compared using the Strict Equality Comparison.</li>
                    <li>Only enumerable "own" properties are considered.</li>
                    <li>Error names and messages are always compared, even if these are not enumerable properties.</li>
                    <li>Enumerable own Symbol properties are compared as well.</li>
                    <li>Object wrappers are compared both as objects and unwrapped values.</li>
                    <li>Object properties are compared unordered.</li>
                    <li>Map keys and Set items are compared unordered.</li>
                    <li>Recursion stops when both sides differ or both sides encounter a circular reference.</li>
                    <li>WeakMap and WeakSet comparison does not rely on their values.</li>
                    <li>If the values are not equal, an AssertionError is thrown with a message property set equal to the value of the message parameter. If the message parameter is undefined, a default error message is assigned. If the message parameter is an instance of an Error then it will be thrown instead of the AssertionError.</li>
                </ul>
                  <li>assert.doesNotReject(block[, error][, message])</li>
                <ul>
                    <li>Added in v10.0.0</li>
                    <ul>
                    <li>block &lt;any&gt;</li>
                    <li>error &lt;RegExp&gt; | &lt;Function&gt;</li>
                    <li>message &lt;any&gt;</li>
                    </ul>
                    <li>Awaits the block promise or, if block is a function, immediately calls the function and awaits the returned promise to complete. It will then check that the promise is not rejected.</li>
                    <li>If block is a function and it throws an error synchronously, assert.doesNotReject() will return a rejected Promise with that error. If the function does not return a promise, assert.doesNotReject() will return a rejected Promise with an ERR_INVALID_RETURN_VALUE error. In both cases the error handler is skipped.</li>
                    <li>Please note: Using assert.doesNotReject() is actually not useful because there is little benefit by catching a rejection and then rejecting it again. Instead, consider adding a comment next to the specific code path that should not reject and keep error messages as expressive as possible.</li>
                    <li>If specified, error can be a Class, RegExp or a validation function.</li>
                    <li>Besides the async nature to await the completion behaves identically to assert.doesNotThrow().</li>
                    <ul>
                    <div class="code-sample">
                        (async () => {<br>
                        <span class="indent">await<br> assert.doesNotReject(<br>
                        async () => {<br>
                            throw new TypeError('Wrong value');</span><br>
                        },<br>
                        SyntaxError<br>
                        );<br>
                        })();<br>
                    </div>
                    </ul>
                    <ul>
                    <div class = "code-sample">
                        assert.doesNotReject(Promise.reject(new TypeError('Wrong value')))<br>
                        <span class = "indent">.then(() => {<br>
                        <span class = "comment">// ...</span><br>
                        });</span><br>
                    </div>
                    </ul>
                </ul>
                <li>assert.doesNotThrow(block[, error][, message])</li>
                <ul>
                    <table border = "1">
                        <tr>
                            <td>Version</td>
                            <td>Changes</td>
                        </tr>
                        <tr>
                            <td>v5.11.0, v4.4.5</td>
                            <td>The message parameter is respected now.</td>
                        </tr>
                        <tr>
                            <td>v4.2.0</td>
                            <td>The error parameter can now be an arrow function.</td>
                        </tr>
                    </table>
                      <ul>
                    <li>block &lt;any&gt;</li>
                    <li>error &lt;RegExp&gt; | &lt;Function&gt;</li>
                    <li>message &lt;any&gt;</li>
                    </ul>
                    <li>Asserts that the function block does not throw an error.</li>
                    <li>Please note: Using assert.doesNotThrow() is actually not useful because there is no benefit by catching an error and then rethrowing it. Instead, consider adding a comment next to the specific code path that should not throw and keep error messages as expressive as possible.</li>
                    <li>When assert.doesNotThrow() is called, it will immediately call the block function.</li>
                    <li>If an error is thrown and it is the same type as that specified by the error parameter, then an AssertionError is thrown. If the error is of a different type, or if the error parameter is undefined, the error is propagated back to the caller.</li>
                    <li>If specified, error can be a Class, RegExp or a validation function. See assert.throws() for more details.</li>
                    <li>The following, for instance, will throw the TypeError because there is no matching error type in the assertion:</li>
                    <ul>
                    <div class = "code-sample">
                        assert.doesNotThrow(<br>
                        <span class = "indent">() => {<br>
                        throw new TypeError('Wrong value');<br>
                        },<br>
                        SyntaxError</span><br>
                        );<br>
                    </div>
                    </ul>
                    <li>However, the following will result in an AssertionError with the message 'Got unwanted exception...':</li>
                    <ul>
                    <div class = "code-sample">
                        assert.doesNotThrow(<br>
                        <span class = "indent">() => {<br>
                        throw new TypeError('Wrong value');<br>
                        },<br>
                        TypeError</span><br>
                        );<br>
                    </div>
                    </ul>
                    <li>If an AssertionError is thrown and a value is provided for the message parameter, the value of message will be appended to the AssertionError message:</li>
                    <ul>
                    <div class = "code-sample">
                        <span class = "indent">assert.doesNotThrow(</span>
                        () => {<br>
                        throw new TypeError('Wrong value');<br>
                        },<br>
                        /Wrong value/,<br>
                        'Whoops'<br>
                        );<br>
                        <span class = "comment">// Throws: AssertionError: Got unwanted exception: Whoops</span>
                    </div>
                    </ul>
                </ul>
                <li>assert.equal(actual, expected[, message])</li>
                <ul>
                    <li>Added in: v0.1.21</li>
                    <ul>
                        <li>actual &lt;any&gt;</li>
                        <li>expected&lt;any&gt;</li>
                        <li>message &lt;any&gt;</li>
                    </ul>
                    <li>Strict Mode</li>
                    <ul>
                        <li>An alias of assert.strictEqual().</li>
                    </ul>
                    <li>Legacy Mode</li>
                    <ul>
                        <li>Tests shallow, coercive equality between the actual and expected parameters using the Abstract Equality Comparison ( == ).</li>
                        <ul>
                        <div class = "code-sample">
                            const assert = require('assert');<br>
                            <br>
                            assert.equal(1, 1);<br>
                            <span class = "comment">// OK, 1 == 1</span><br>
                            assert.equal(1, '1');<br>
                            <span class = "comment">// OK, 1 == '1'</span><br>
                            <br>
                            assert.equal(1, 2);<br>
                            <span class = "comment">// AssertionError: 1 == 2</span><br>
                            assert.equal({ a: { b: 1 } }, { a: { b: 1 } });<br>
                            <span class = "comment">// AssertionError: { a: { b: 1 } } == { a: { b: 1 } }</span><br>
                        </div>
                        </ul>
                        <li>If the values are not equal, an AssertionError is thrown with a message property set equal to the value of the message parameter. If the message parameter is undefined, a default error message is assigned. If the message parameter is an instance of an Error then it will be thrown instead of the AssertionError.</li>
                    </ul>
                </ul>
                <li>assert.fail([message])</li>
                <ul>
                    <li>Added in v0.1.21</li>
                    <ul>
                        <li>message &lt;any&gt; Default: 'Failed'</li>
                    </ul>
                    <li>Throws an AssertionError with the provided error message or a default error message. If the message parameter is an instance of an Error then it will be thrown instead of the AssertionError.</li>
                    <ul>
                    <div class = "code-sample">
                        const assert = require('assert').strict;<br>
                        <br>
                        assert.fail();<br>
                        <span class = "comment">// AssertionError [ERR_ASSERTION]: Failed<br></span>
                        <br>
                        assert.fail('boom');<br>
                        <span class = "comment">// AssertionError [ERR_ASSERTION]: boom<br></span>
                        <br>
                        assert.fail(new TypeError('need array'));<br>
                        <span class = "comment">// TypeError: need array<br></span>
                    </div>
                    </ul>
                </ul>
                <li>assert.fail(actual, expected[, message[, operator[, stackStartFunction]]])</li>
                <ul>
                <table border = "1">
                    <tr>
                        <td>Version</td>
                        <td>Changes</td>
                    </tr>
                    <tr>
                        <td>v10.0.0</td>
                        <td>Calling assert.fail() with more than one argument is deprecated and emits a warning.</td>
                    </tr>
                    <tr>
                        <td>v0.1.21</td>
                        <td>Added in: v0.1.21</td>
                    </tr>
                </table>
                <ul>
                    <li>actual &lt;any&gt;</li>
                    <li>expected&lt;any&gt;</li>
                    <li>message &lt;any&gt;</li>
                    <li>operator &lt;string&gt; <b>Default:</b> '!='</li>
                    <li>StackStartFunction &lt;Function&gt; <b>Default:</b> assert.fail</li>
                </ul>
                <li>If message is falsy, the error message is set as the values of actual and expected separated by the provided operator. If just the two actual and expected arguments are provided, operator will default to '!='. If message is provided as third argument it will be used as the error message and the other arguments will be stored as properties on the thrown object. If stackStartFunction is provided, all stack frames above that function will be removed from stacktrace (see Error.captureStackTrace). If no arguments are given, the default message Failed will be used.</li>
                <ul>
                <div class = "code-sample">
                    const assert = require('assert').strict;<br>
                    <br>
                    assert.fail('a', 'b');<br>
                    <span class = "comment">// AssertionError [ERR_ASSERTION]: 'a' != 'b'</span><br>
                    <br>
                    assert.fail(1, 2, undefined, '>');<br>
                    <span class = "comment">// AssertionError [ERR_ASSERTION]: 1 > 2</span><br>
                    <br>
                    assert.fail(1, 2, 'fail');<br>
                    <span class = "comment">// AssertionError [ERR_ASSERTION]: fail</span><br>
                    <br>
                    assert.fail(1, 2, 'whoops', '>');<br>
                    <span class = "comment">// AssertionError [ERR_ASSERTION]: whoops</span><br>
                    <br>
                    assert.fail(1, 2, new TypeError('need array'));<br>
                    <span class = "comment">// TypeError: need array</span><br>
                </div>
                </ul>
                    <li>In the last three cases actual, expected, and operator have no influence on the error message.</li>
                    <li>Example use of stackStartFunction for truncating the exception's stacktrace:</li>
                    <ul>
                        <div class = "code-sample">
                            function suppressFrame() {<br>
                            <span class = "indent">assert.fail('a', 'b', undefined, '!==', suppressFrame);</span><br>
                            }<br>
                            suppressFrame();<br>
                            <span class = "comment">// AssertionError [ERR_ASSERTION]: 'a' !== 'b'</span><br>
                            <span class = "comment">//     at repl:1:1<br></span>
                            <span class = "comment">//     at ContextifyScript.Script.runInThisContext (vm.js:44:33)<br></span>
                            <span class = "comment">//     ...<br></span>
                        </div>
                    </ul>
                </ul>
                <li>assert.ifError(value)</li>
                <ul>
                    <table border = "none">
                        <tr>
                            <td>History</td>
                            <td>Changes</td>
                        </tr>
                        <tr>
                            <td>v10.0.0</td>
                            <td>Instead of throwing the original error it is now wrapped into an AssertionError that contains the full stack trace.</td>
                        </tr>
                        <tr>
                            <td>v10.0.0</td>
                            <td>Value may now only be undefined or null. Before any truthy input was accepted.</td>
                        </tr>
                        <tr>
                            <td>v0.1.97</td>
                            <td>Added in: v0.1.97</td>
                        </tr>
                    </table>
                    <ul>
                        <li>value &lt;any&gt;</li>
                        <ul>
                        <li>Throws value if value is not undefined or null. This is useful when testing the error argument in callbacks. The stack trace contains all frames from the error passed to ifError() including the potential new frames for ifError() itself.</li>
                    </ul>
                    </ul>
                </ul>
                <li>assert.notDeepEqual(actual, expected[, message])</li>
                <ul>
                    <table border = "1">
                        <tr>
                            <td>Version</td>
                            <td>Changes</td>
                        </tr>
                        <tr>
                            <td>v9.0.0</td>
                            <td>The Error names and messages are now properly compared</td>
                        </tr>
                        <tr>
                            <td>v8.0.0</td>
                            <td>The Set and Map content is also compared</td>
                        </tr>
                        <tr>
                            <td>v6.4.0, v4.7.1</td>
                            <td>Typed array slices are handled correctly now.</td>
                        </tr>
                        <tr>
                            <td>v6.1.0, v4.5.0</td>
                            <td>Objects with circular references can be used as inputs now.</td>
                        </tr>
                        <tr>
                            <td>v5.10.1, v4.4.3</td>
                            <td>Handle non-Uint8Array typed arrays correctly.</td>
                        </tr>
                        <tr>
                            <td>v0.1.21</td>
                            <td>Added in: v0.1.21</td>
                        </tr>
                    </table>
                    <ul>
                        <li>actual &lt;any&gt;</li>
                        <li>expected&lt;any&gt;</li>
                        <li>message &lt;any&gt;</li>
                    </ul>
                    <li>Strict Mode</li>
                    <ul>
                        <li>An alias of assert.notDeepStrictEqual().</li>
                    </ul>
                    <li>Legacy Mode</li>
                    <ul>
                        <li>If the values are deeply equal, an AssertionError is thrown with a message property set equal to the value of the message parameter. If the message parameter is undefined, a default error message is assigned. If the message parameter is an instance of an Error then it will be thrown instead of the AssertionError.</li>
                    </ul>
                </ul>
                <li>assert.notDeepStrictEqual(actual, expected[, message])</li>
                <ul>
                    <table border = "1">
                        <tr>
                            <td>History</td>
                            <td>Changes</td>
                        </tr>
                        <tr>
                            <td>v9.0.0</td>
                            <td>The -0 and +0 are not considered equal anymore.</td>
                        </tr>
                        <tr>
                            <td>v9.0.0</td>
                            <td>The NaN is now compared using the SameValueZero comparison.</td>
                        </tr>
                        <tr>
                            <td>v9.0.0</td>
                            <td>The Error names and messages are now properly compared</td>
                        </tr>
                        <tr>
                            <td>v8.0.0</td>
                            <td>The Set and Map content is also compared</td>
                        </tr>
                        <tr>
                            <td>v6.4.0, v4.7.1</td>
                            <td>Typed array slices are handled correctly now.</td>
                        </tr>
                        <tr>
                            <td>v6.1.0</td>
                            <td>Objects with circular references can be used as inputs now.</td>
                        </tr>
                        <tr>
                            <td>v5.10.1, v4.4.3</td>
                            <td>Handle non-Uint8Array typed arrays correctly.</td>
                        </tr>
                        <tr>
                            <td>v1.2.0</td>
                            <td>Added in: v1.2.0</td>
                        </tr>
                    </table>
                    <ul>
                        <li>actual &lt;any&gt;</li>
                        <li>expected&lt;any&gt;</li>
                        <li>message &lt;any&gt;</li>
                    </ul>
                    <li>Tests for deep strict inequality. Opposite of assert.deepStrictEqual().</li>
                    <ul>
                        <div class = "code-sample">
                            const assert = require('assert').strict;<br>
                            <br>
                            assert.notDeepStrictEqual({ a: 1 }, { a: '1' });<br>
                            <span class = "comment">// OK</span>
                        </div>
                    </ul>
                    <li>If the values are deeply and strictly equal, an AssertionError is thrown with a message property set equal to the value of the message parameter. If the message parameter is undefined, a default error message is assigned. If the message parameter is an instance of an Error then it will be thrown instead of the AssertionError.</li>
                </ul>
                <li>assert.notEqual(actual, expected[, message])</li>
                <ul>
                    <ul>
                        <li>actual &lt;any&gt;</li>
                        <li>expected&lt;any&gt;</li>
                        <li>message &lt;any&gt;</li>
                    </ul>
                    <li>Strict Mode</li>
                    <ul>
                        <li>An alias of assert.notStrictEqual().</li>
                    </ul>
                    <li>Legacy Mode</li>
                    <ul>
                        <li>Tests shallow, coercive inequality with the Abstract Equality Comparison ( != ).</li>
                        <ul>
                            <div class = "code-sample">
                                const assert = require('assert');<br>
                                <br>
                                assert.notEqual(1, 2);<br>
                                <span class = "comment">// OK<br></span>
                                <br>
                                assert.notEqual(1, 1);<br>
                                <span class = "comment">// AssertionError: 1 != 1</span>
                                <br>
                                assert.notEqual(1, '1');<br>
                                <span class = "comment">// AssertionError: 1 != '1'</span><br>
                            </div>
                        </ul>
=======
                        ]</li>
                        <li>Legacy Components</li>
                        <ul>
                           <li>The Abstract Equality Comparison will be used for any function without "strict" in its name, such as assert.deepEqual().</li>
>>>>>>> afeb0f9ac4833b7b41d7e1941c4df77f5d85e1db
                    </ul>
                </ul>
            </ul>
        </ul>
        </div>
    </body>
</html>
